src/server/bg_services/replication_log_parser.js

```diff
@@ -18,19 +18,22 @@ const moment = require('moment');
  * @param {Record<any, any>} replicationconfig Replication configuration
  * @param {Number} candidates_limit Candidates limit
  * @param {(log_entry: Record<string, any>) => boolean} ignore_fn - function that returns true if the log entry should be ignored
- * @returns 
+ * @returns {Promise<{
+ *  items: nb.ReplicationLogCandidates,
+ *  done: () => Promise<void>
+ * }>} Candidates
  */
 async function get_log_candidates(source_bucket_id, rule_id, replicationconfig, candidates_limit, ignore_fn = () => false) {
     return get_aws_log_candidates(source_bucket_id, rule_id, replicationconfig, candidates_limit, ignore_fn);
 }
 
 async function get_aws_log_candidates(source_bucket_id, rule_id, replication_config, candidates_limit, ignore_fn) {
-    let aws_log_replication_info = replication_config.log_replication_info.aws_log_replication_info;
+    const aws_log_replication_info = replication_config.log_replication_info.aws_log_replication_info;
     const { logs_bucket, prefix } = aws_log_replication_info.logs_location;
-    let continuation_token = get_continuation_token_for_rule(rule_id, replication_config);
     const s3 = get_source_bucket_aws_connection(source_bucket_id, aws_log_replication_info);
+    let continuation_token = get_continuation_token_for_rule(rule_id, replication_config);
 
-    let candidates_obj = {};
+    const candidates = [];
```

Comment:
Why do you keep using an array approach instead of having an object of the following form - { key: [{time: Date, action: string] }, sorting the inner arrays and then skipping situations like put,put,delete that happened in different timestamps.